openid-4-verifiable-presentations-1_0.md

```diff
@@ -542,11 +548,250 @@ The Wallet then validates the request as specified in OAuth 2.0 [@RFC6749].
+Verifiable Credentials, as defined in (#credential_query).
+
+Note: While this specification does not define additional top-level properties,
+future extensions MAY define additional properties that modify the behavior of
+the Wallet when processing the query.
+
+## Credential Query {#credential_query}
+
+A Credential Query is an object representing a request for one specific
+credential. It contains the following properties:
+
+`group`:
+: REQUIRED. A string identifying the credential in the response, or, if multiple
+credentials carry the same group identifier, the group of credentials that the
+requested credential belongs to.
+
+`format`:
+: REQUIRED. A string that specifies the format of the requested
+Verifiable Credential. Valid Credential format identifier values are defined in
+Appendix A of [@!OpenID.VCI]. The value of this property MUST be one of the
+supported Credential formats as defined in the Wallet's metadata.
+
+`expect_meta`: 
+: OPTIONAL. An object defining additional properties requested by the Verifier that
+apply to the metadata and validity data of the credential. The properties of
+this object are defined per format in (#format_specific_properties). If omitted,
+no specific restrictions are placed on the metadata or validity of the requested
+Credential.
+
+`expect_claims`:
+: OPTIONAL. An array that specifies the claims that the requested
+Verifiable Credential must contain. Each entry in the array corresponds to exactly
+one claim that is to be returned by the Wallet to fulfill the credential
+request. 
+
+`alternative_credentials`:
+: OPTIONAL. An array of credential group identifiers that specify alternative
+groups of credentials that MAY be returned instead of the credential. Any credential
+in a group listed in `alternative_credentials` MUST NOT itself contain an
+`alternative_credentials` property.
+
+Each entry MUST be an object with the following properties:
+
+`group`:
+: OPTIONAL. A string identifying a group of claims for expressing alternatives
+between claims. Multiple claims MAY have the same group identifier. The group
+identifier for claims is not reflected in the response.
+
+`path`:
+: REQUIRED if the credential format uses a JSON-based claims structure; MUST NOT
+be present otherwise. The value MUST be a claims path pointer that specifies the path to the claim
+within the Verifiable Credential, as defined in (#claims_path_pointer).
+
+`namespace`:
+: REQUIRED if the credential format is based on ISO 18013-5; MUST NOT be present otherwise. 
+The value MUST be a string that specifies the namespace of the claim
+within the Verifiable Credential, e.g., `org.iso.18013.5.1`.
+
+`claim_name`:
+: REQUIRED if the credential format is based on ISO 18013-5; MUST NOT be present otherwise. 
+The value MUST be a string that specifies the name of the claim within the provided namespace
+in the Verifiable Credential, e.g., `first_name`.
+
+`value`:
```

Comment:
Matching of values is a tricky subject. Some of the issues:

- Matching on values can be complicated for wallet / credential selection. Because you may not have access to the value of the element at the selection time.
- Even after wallet selection, having access to the values of the data elements may still only be available after user consent because of security reasons (decryption etc). Meaning that the presentation may fail only after user consent.
- Values of data elements are not just strings, integers or booleans. Depending on the credential format they can be maps, arrays, byte strings  or other items.
 
So besides requiring a stricter definition, mostly it's a mechanism that may just not be available because of security or ux reasons, so I would not be in favor of having it as an option since it's not a mechanism RPs can rely on.

Reply:
It is a tricky subject, but the working group did spent a significant amount of time analysing what is needed from a query language for various use cases, and the conclusion was that matching of values is something that must be supported by a query language: https://github.com/openid/OpenID4VP/issues/144

It's also something supported by OID4VP today in presentation exchange, so something that is likely to be required if we want to move back to having only one query language in the future.

I acknowledge the difficulties, but I think we need to focus on how it can be done (and not focus on whether it should be done or not). Whether it's something that's mandatory to implement or not is I think something that's more likely to be dictated by the ecosystems. I understand some of the mandated EU use cases will require matching.

Reply:
We have syntax for other types of matching in the [Advanced Syntax for Claims specification](https://openid.bitbucket.io/ekyc/openid-connect-advanced-syntax-for-claims.html#name-defining-transformed-claims) which we could re-use here. There is at least one implementation of that spec, so we're not starting from scratch.

Reply:
Please see a proposal here: https://github.com/openid/OpenID4VP/pull/266#issuecomment-2362093202